File: lintcode/python/0928_longest_substring_with_at_most_two_distinct_characters.py
#!/usr/bin/python -t

# two pointers

class Solution:
    """
    @param s: a string
    @return: the length of the longest substring T that contains at most 2 distinct characters
    """
    def lengthOfLongestSubstringTwoDistinct(self, s):
        # Write your code here
        n = len(s)
        d = {}
        left = 0
        ret = 0
        
        for right in range(n):
            d[s[right]] = d.get(s[right], 0) + 1
            
            while left <= right and len(d) > 2:
                d[s[left]] -= 1
                if d[s[left]] == 0:
                    del d[s[left]]
                    
                left += 1
                
            ret = max(ret, right-left+1)
            
        return ret
        
     

File: lintcode/golang/0928_longest_substring_with_at_most_two_distinct_characters.go

// hash and two pointers

/**
 * @param s: a string
 * @return: the length of the longest substring T that contains at most 2 distinct characters
 */
func lengthOfLongestSubstringTwoDistinct (s string) int {
    // Write your code here
    n := len(s)
    if n <= 2 {
        return n
    }

    ret := 0
    l := 0
    d := make(map[byte]int)

    for r := 0; r < n; r++ {
        d[s[r]]++

        for len(d) > 2 {
            d[s[l]]--
            if d[s[l]] == 0 {
                delete(d, s[l])
            }
            l++
        }

        if len(d) == 2 {
            ret = max(ret, r-l+1)
        }
    }

    return ret
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}


/**
 * @param s: a string
 * @return: the length of the longest substring T that contains at most 2 distinct characters
 */
func lengthOfLongestSubstringTwoDistinct (s string) int {
    // Write your code here
    m := make(map[byte]int)
    n := len(s)
    if n <= 2 {
        return n
    }
    
    l := 0
    ret := 0
    
    for r := 0; r < n; r++ {
        m[s[r]] += 1
        
        for l <= r && len(m) > 2 {
            m[s[l]]--
            if m[s[l]] == 0 {
                delete(m, s[l])
            }
            l++
        }
        
        ret = max(ret, r-l+1)
    }
    
    return ret
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
